python/lib/communication/dmod/communication/client.py

```diff
@@ -435,225 +662,180 @@ async def __aexit__(self, *exc_info):
 
     async def async_recv(self) -> Union[str, bytes]:
         """
-        Receive data over the websocket connection.
+        Receive data over the connection.
 
         Returns
         -------
         Union[str, bytes]
             The data received over the connection.
         """
-        with self as websocket:
-            return await websocket.connection.recv()
+        with self as connection_owner:
+            return await connection_owner._connection_recv()
 
     @property
-    def client_ssl_context(self) -> ssl.SSLContext:
-        """
-        Get the client SSL context property, lazily instantiating if necessary.
-
-        Returns
-        -------
-        ssl.SSLContext
-            the client SSL context for secure connections
-        """
-        if self._client_ssl_context is None:
-            self._client_ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
-            endpoint_pem = self._ssl_directory.joinpath(self._cert_pem_file_basename)
-            self.client_ssl_context.load_verify_locations(endpoint_pem)
-        return self._client_ssl_context
+    def connection(self) -> Optional[CONN]:
+        return self._connection
 
 
-class InternalServiceClient(WebSocketClient, Generic[M, R], ABC):
+class WebSocketClient(SSLSecuredTransportLayerClient, ConnectionContextClient[websockets.WebSocketClientProtocol]):
     """
-    Abstraction for a client that interacts with some internal, non-public-facing DMOD service.
+    Subtype of ::class:`SSLSecuredTransportLayerClient` that specifically works over SSL-secured websocket connections.
+
+    A websocket-based implementation of ::class:`SSLSecuredTransportLayerClient`.  Instances are also async context
+    managers for runtime contexts that handle websocket connections, with the manager function returning the instance
+    itself.
+
+    A new runtime context will check whether there is an open websocket connection already and open a connection if not.
+    In all cases, it maintains an instance attribute that is a counter of the number of active usages of the connection
+    (i.e., the number of separate, active contexts).  When the context is exited, the instance's active usage counter is
+    reduced by one and, if that context represents the last active use of the connection, the connection object is
+    closed and then has its reference removed.
+
+    The ::method:`async_send` and ::method:`async_recv` functions can be used without already being in an active context
+    (i.e., they will enter a new context for the scope of the function). However, within in an already open context,
+    calls to ::method:`async_send` and ::method:`async_recv` can be used as needed to support arbitrarily communication
+    over the websocket.
     """
 
     @classmethod
-    @abstractmethod
-    def get_response_subtype(cls) -> Type[R]:
+    def build_endpoint_uri(cls, host: str, port: Union[int, str], path: Optional[str] = None, is_secure: bool = True):
```

Comment:
Should this instead just be a static method or just a function?

Reply:
As things are right now, a static method would probably be slightly more appropriate.  

This seems like something we may tweak in the future once we have more transport client implementations.  That probably has something to do with why it was already a class method.  I'll adjust it now, though, and we can always change it back later.